File: frontend/src/types/game.ts
// TypeScript types for WSIM game models
// These mirror the Pydantic models in backend/wsim_core/models

export type Side = "P1" | "P2";

export type Facing = "N" | "NE" | "E" | "SE" | "S" | "SW" | "W" | "NW";

export type WindDirection = "N" | "NE" | "E" | "SE" | "S" | "SW" | "W" | "NW";

export type LoadState = "E" | "R"; // E = Empty, R = Roundshot

export type GamePhase = "planning" | "movement" | "combat" | "reload";

export type Broadside = "L" | "R"; // L = Left, R = Right

export type AimPoint = "hull" | "rigging";

export interface HexCoord {
  col: number;
  row: number;
}

export interface Ship {
  id: string;
  name: string;
  side: Side;
  bow_hex: HexCoord;
  stern_hex: HexCoord;
  facing: Facing;
  battle_sail_speed: number;
  guns_L: number;
  guns_R: number;
  carronades_L: number;
  carronades_R: number;
  hull: number;
  rigging: number;
  crew: number;
  marines: number;
  load_L: LoadState;
  load_R: LoadState;
  fouled: boolean;
  struck: boolean;
  turns_without_bow_advance: number;
}

export interface ShipOrders {
  ship_id: string;
  movement: string;
}

export interface TurnOrders {
  turn_number: number;
  side: Side;
  orders: ShipOrders[];
  ready: boolean;
}

export interface DiceRoll {
  num_dice: number;
  die_type: number;
  rolls: number[];
  total: number;
}

export interface EventLogEntry {
  turn_number: number;
  phase: GamePhase;
  event_type: string;
  summary: string;
  dice_roll?: DiceRoll;
  modifiers?: Record<string, number>;
  state_diff?: Record<string, unknown>;
  metadata?: Record<string, unknown>;
}

export interface Game {
  id: string;
  scenario_id: string;
  turn_number: number;
  phase: GamePhase;
  map_width: number;
  map_height: number;
  wind_direction: WindDirection;
  ships: Record<string, Ship>;
  p1_orders: TurnOrders | null;
  p2_orders: TurnOrders | null;
  event_log: EventLogEntry[];
  turn_limit: number | null;
  victory_condition: string;
}

// Scenario types
export interface MapConfig {
  width: number;
  height: number;
}

export interface WindConfig {
  direction: WindDirection;
}

export interface VictoryConfig {
  type: string;
  metric?: string;
}

export interface ShipStartPosition {
  bow: [number, number];
  facing: Facing;
}

export interface ShipGuns {
  L: number;
  R: number;
}

export interface ShipCarronades {
  L: number;
  R: number;
}

export interface ShipInitialLoad {
  L: LoadState;
  R: LoadState;
}

export interface ScenarioShip {
  id: string;
  side: Side;
  name: string;
  battle_sail_speed: number;
  start: ShipStartPosition;
  guns: ShipGuns;
  carronades: ShipCarronades;
  hull: number;
  rigging: number;
  crew: number;
  marines: number;
  initial_load: ShipInitialLoad;
}

export interface Scenario {
  id: string;
  name: string;
  description: string;
  map: MapConfig;
  wind: WindConfig;
  turn_limit: number | null;
  victory: VictoryConfig;
  ships: ScenarioShip[];
}

export interface ScenarioInfo {
  id: string;
  name: string;
  description: string;
}

// API request/response types
export interface CreateGameRequest {
  scenario_id: string;
}

export interface CreateGameResponse {
  game_id: string;
  state: Game;
}

export interface SubmitOrdersRequest {
  side: Side;
  orders: ShipOrders[];
}

export interface SubmitOrdersResponse {
  state: Game;
  orders_submitted: boolean;
}

export interface MarkReadyRequest {
  side: Side;
}

export interface MarkReadyResponse {
  state: Game;
  ready: boolean;
  both_ready: boolean;
}

export interface AdvanceTurnResponse {
  state: Game;
}

export interface ResolvePhaseResponse {
  state: Game;
  events: EventLogEntry[];
}

export interface FireBroadsideRequest {
  ship_id: string;
  broadside: Broadside;
  target_ship_id: string;
  aim: AimPoint;
}

export interface FireBroadsideResponse {
  state: Game;
  events: EventLogEntry[];
}

File: backend/wsim_core/models/game.py
"""Game state model."""

from pydantic import BaseModel, Field

from .common import GamePhase, WindDirection
from .events import EventLogEntry
from .orders import TurnOrders
from .ship import Ship


class Game(BaseModel):
    """Complete game state.

    Represents the entire state of a game including ships, wind, phase, and history.
    """

    # Identity
    id: str = Field(description="Unique game identifier")
    scenario_id: str = Field(description="Scenario this game is based on")

    # Game state
    turn_number: int = Field(ge=1, default=1, description="Current turn number")
    phase: GamePhase = Field(default=GamePhase.PLANNING, description="Current game phase")

    # Map and environment
    map_width: int = Field(ge=1, description="Map width in hexes")
    map_height: int = Field(ge=1, description="Map height in hexes")
    wind_direction: WindDirection = Field(description="Current wind direction")

    # Ships
    ships: dict[str, Ship] = Field(description="All ships indexed by ship_id")

    # Current turn orders
    p1_orders: TurnOrders | None = Field(default=None, description="P1 orders for current turn")
    p2_orders: TurnOrders | None = Field(default=None, description="P2 orders for current turn")

    # Event log
    event_log: list[EventLogEntry] = Field(
        default_factory=list, description="Complete event history"
    )

    # Victory conditions
    turn_limit: int | None = Field(default=None, description="Maximum turns (None = unlimited)")
    victory_condition: str = Field(default="first_struck", description="Victory condition type")
    game_ended: bool = Field(default=False, description="Whether the game has ended")
    winner: str | None = Field(default=None, description="Winner side (P1, P2, or None for draw)")

    def get_ship(self, ship_id: str) -> Ship:
        """Get a ship by ID.

        Args:
            ship_id: The ship identifier

        Returns:
            The ship

        Raises:
            KeyError: If ship doesn't exist
        """
        return self.ships[ship_id]

    def get_ships_by_side(self, side: str) -> list[Ship]:
        """Get all ships for a given side.

        Args:
            side: The player side (P1 or P2)

        Returns:
            List of ships for that side
        """
        return [ship for ship in self.ships.values() if ship.side == side]

    def add_event(self, event: EventLogEntry) -> None:
        """Add an event to the log.

        Args:
            event: The event to add
        """
        self.event_log.append(event)
